wandb/integration/ultralytics/bbox_utils.py

```diff
@@ -0,0 +1,194 @@
+    for prediction in model_predictions:
+        pred_metadata_map.setdefault(prediction["image_id"], [])
+        pred_metadata_map[prediction["image_id"]].append(prediction)
+
+    return pred_metadata_map
+
+
+def get_mean_confidence_map(
+    classes: List, confidence: List, class_id_to_label: Dict
+) -> Dict:
+    confidence_map = {v: [] for _, v in class_id_to_label.items()}
+    for class_idx, confidence_value in zip(classes, confidence):
+        confidence_map[class_id_to_label[class_idx]].append(confidence_value)
+    for label, confidence_list in confidence_map.items():
+        if len(confidence_list) > 0:
+            confidence_map[label] = sum(confidence_list) / len(confidence_list)
+        else:
+            confidence_map[label] = 0
+    return confidence_map
+
+
+def get_boxes(result: Results) -> Tuple[Dict, Dict]:
+    boxes = result.boxes.xywh.long().numpy()
+    classes = result.boxes.cls.long().numpy()
+    confidence = result.boxes.conf.numpy()
+    class_id_to_label = {int(k): str(v) for k, v in result.names.items()}
+    mean_confidence_map = get_mean_confidence_map(
+        classes, confidence, class_id_to_label
+    )
+    box_data, total_confidence = [], 0.0
+    for idx in range(len(boxes)):
+        box_data.append(
+            {
+                "position": {
+                    "middle": [int(boxes[idx][0]), int(boxes[idx][1])],
+                    "width": int(boxes[idx][2]),
+                    "height": int(boxes[idx][3]),
+                },
+                "domain": "pixel",
+                "class_id": int(classes[idx]),
+                "box_caption": class_id_to_label[int(classes[idx])],
+                "scores": {"confidence": float(confidence[idx])},
+            }
+        )
+        total_confidence += float(confidence[idx])
+    boxes = {
+        "predictions": {
+            "box_data": box_data,
+            "class_labels": class_id_to_label,
+        },
+    }
+    return boxes, mean_confidence_map
+
+
+def plot_predictions(
+    result: Results, table: Optional[wandb.Table] = None
+) -> Union[wandb.Table, Tuple[wandb.Image, Dict, Dict]]:
+    result = result.to("cpu")
+    boxes, mean_confidence_map = get_boxes(result)
+    image = wandb.Image(result.orig_img[:, :, ::-1], boxes=boxes)
+    if table is not None:
+        table.add_data(
+            image,
+            len(boxes["predictions"]["box_data"]),
+            mean_confidence_map,
+            result.speed,
+        )
+        return table
+    return image, boxes["predictions"], mean_confidence_map
+
+
+def plot_validation_results(
+    dataloader,
+    class_label_map,
+    predictor: DetectionPredictor,
+    table: wandb.Table,
+    max_validation_batches: int,
+    epoch: Optional[int] = None,
+) -> wandb.Table:
+    data_idx = 0
+    for batch_idx, batch in enumerate(dataloader):
+        for img_idx, image_path in enumerate(batch["im_file"]):
+            prediction_result = predictor(image_path)[0]
```

Comment:
Shouldn't we do batched predictions and convert the results for a batch instead ? Won't predicting on each image increase the latency of the callback during plotting as a whole ?